src/libcore/array/mod.rs

```diff
@@ -387,6 +387,84 @@ where
     }
 }
 
+/// Non-zero numbers
+///
+/// We require following from this trait:
+///  - [(); 0] does **not** implement NonZero.
+///  - [(); N] where N != 0 **implements** NonZero.
+///  As side effect of selected approach, this trait is also implemented
+///  for all other types, but it does not matter.
+#[unstable(feature = "array_default_internals", reason = "implementation detail", issue = "none")]
+pub auto trait NonZero {}
+#[unstable(feature = "array_default_internals", reason = "implementation detail", issue = "none")]
+impl !NonZero for [(); 0] {}
```

Comment:
You can avoid exposing this publicly by "sealing" it, i.e. putting in a private `mod sealed {...}` (can be inline, don't need a separate file) and referring to it as `sealed::NonZero` from the rest of the module.

Might be able to remove `#[unstable]` afterwards although I'm not 100% sure that logic relies on reachability.